src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -50,69 +54,127 @@
 
     protected abstract List<FieldSignature> staticFieldSignatures();
 
-    // TODO should precompute hash sets of signatures, assuming we document that the signatures may not change over the lifetime of the whitelist (or pass them in the constructor)
+    ConcurrentHashMap<String, Boolean> permittedCache = new ConcurrentHashMap<String, Boolean>();  // Not private to facilitate testing
+
+    private void cacheSignatureList(List<Signature> ...sigs) {
+        for (List<Signature> list : sigs) {
+            for (Signature s : list) {
+                if (!s.isWildcard()) { // Cache entries for wildcard signatures will never be accessed and just waste space
+                    permittedCache.put(s.toString(), Boolean.TRUE);
+                }
+            }
+        }
+    }
+
+    /** Prepopulates the "permitted" cache. */
+    public void precache() {
+        cacheSignatureList((List)methodSignatures(), (List)(newSignatures()),
+                           (List)(staticMethodSignatures()), (List)(fieldSignatures()),
+                           (List)(staticFieldSignatures()));
+    }
+
+    /** Frees up memory used for the cache. */
+    void clearCache() {
+        this.permittedCache.clear();
+        this.permittedCache = new ConcurrentHashMap<String, Boolean>();
+    }
 
     @Override public final boolean permitsMethod(Method method, Object receiver, Object[] args) {
+        String key = Whitelist.canonicalMethodSig(method);
+        Boolean b = permittedCache.get(key);
+        if (b != null) {
+            return b;
+        }
+
+        boolean output = false;
         for (MethodSignature s : methodSignatures()) {
             if (s.matches(method)) {
-                return true;
+                output = true;
+                break;
             }
         }
-        return false;
+        permittedCache.put(key, output);
+        return output;
```

Comment:
This makes no sense to me. If we already precached all the signatures, why are we even iterating them again now? If it is not in the cache, just `return false` right away. Then we could delete the `matches` methods too, and make `permittedCache` be a `Set<String>`.

I guess you are doing this just to handle wildcard signatures? But as I noted in #179, that could be more easily (and efficiently) handled by precaching wildcards (i.e., deleting `isWildcard` and its call in `cacheSignatureList`) and having `permitsMethod` just do something like

```java
return permittedCache.contains(Whitelist.canonicalMethodSig(method)) ||
    permittedCache.contains(Whitelist.canonicalMethodSigStar(method));
```

where `canonicalMethodSigStar` just adds `*` rather than `method.getName()`. Analogously for `permitsField`.

Reply:
Yes, it's to deal with wildcards.  We could indeed double-probe using a wildcard signature -- more elegant, but this implementation is already massively improved vs. the previous.

Can always be addressed later if needed -- I doubt the performance difference will matter much, since we're already caching FALSE.

Reply:
> I doubt the performance difference will matter much

Agreed, was suggesting this more to simplify the implementation. I care more about that in this code area than others because of the possible security implications of mistakes.